sqlmesh/core/plan/definition.py

```diff
@@ -414,28 +414,30 @@ def _missing_intervals(self) -> t.Dict[t.Tuple[str, str], Intervals]:
             # to choose non-breaking / forward only. this will change the version of the snapshot on the fly
             # thus changing the missing intervals. additionally we replace any snapshots with the old copies
             # because they have intervals and the ephemeral ones don't
-            snapshots = {
-                (snapshot.name, snapshot.version_get_or_generate()): snapshot
-                for snapshot in self.snapshots
+            old_snapshots = {
+                (old.name, old.version_get_or_generate()): old
+                for _, old in self.context_diff.modified_snapshots.values()
             }
```

Comment:
why did this change from snapshots to modified_snapshots

Reply:
because the original logic didn't make much sense. We're only missing intervals for newly created snapshots and that's what we should hydrate instead of replacing new snapshots with old ones.

Reply:
is the comment above still valid? i remember this section was quite tricky and i added this in order to fix a very tricky bug with changing categorizations